### PR TITLE
Create the proper shortcode on paste

### LINKED ISSUE #21103: Shortcode transforms on paste no longer work
**Describe the bug**
When pasting a pre-made shortcode string into the paragraph block it no longer transforms into the block with the relevant transform defined. 

I.e pasting in `[gallery ids="13,12,14"]` does not transform into a gallery block even though the transform is defined for it: https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/gallery/transforms.js#L60

It isn't even converted into the generic shortcode.

**To reproduce**
Steps to reproduce the behaviour:
1. Create a new post
2. Paste the following into a paragraph block: `[gallery ids="13,12,14"]` 
3. Note that it remains as plain text in the paragraph block and that it isn't transformed into the gallery block

**Expected behaviour**
Pasting a pre-made shortcode into the paragraph block should transform it into the relevant block that has a transform defined for it. It used to do so but we're not sure when they stopped working.

**Additional context**
- This is happening on 5.3.2 and on the various 5.4-RC3 nightlies we've been testing
- I've replicated the issue on a clean install of WordPress with the twentytwenty theme installed
- Its happening for users on Mac and PC and on Firefox and Chrome
- If I create a classic block, put the gallery shortcode into that and then `Convert to blocks` it is correctly transformed to a gallery block.

### PR DESCRIPTION
**Pull request by oandregal**
Fixes https://github.com/WordPress/gutenberg/issues/21103

## How to test

- Create a new post
- Paste the following into a paragraph block: `[gallery ids="13,12,14"]` (change these ids for ids of elements in your media library).
- Expected result is that a gallery block should be created.

### CODE REVIEW

File: packages/blocks/src/api/raw-handling/paste-handler.js
Comment (nosolosw) [inline packages/blocks/src/api/raw-handling/paste-handler.js:208]: It seems to me that this should be fine because shortcodes are converted to blocks, which can't be inline. cc @ellatrix

**Comment by nosolosw** [comment]
cc @hrkhal you may be interested in testing this one.

File: packages/blocks/src/api/raw-handling/paste-handler.js
Comment (nosolosw) [reply packages/blocks/src/api/raw-handling/paste-handler.js:208]: I've only checked the [default shortcodes](https://codex.wordpress.org/Shortcode_API#Default_Shortcodes), though. Perhaps, there is some shortcode that could be converted into inline formats. However, I don't see that's possible to do right now with the transformations API.

Comment (ellatrix) [reply packages/blocks/src/api/raw-handling/paste-handler.js:208]: Yeah, I think it's fine if all the tests pass.

**Review by ellatrix — APPROVED** [review]

**Comment by ellatrix** [comment]
Maybe it's good to have @mcsf's opinion as well, since you did some work on this. :)

**Comment by mcsf** [comment]
I can't recall: were there specific Puppeteer issues keeping from having e2e tests for pasting? It could be nice to add some.

File: packages/blocks/src/api/raw-handling/paste-handler.js
Comment (mcsf) [reply packages/blocks/src/api/raw-handling/paste-handler.js:208]: I disagree, I think it's fairly important to respect any third-party **inline** shortcodes. In my experience, the use cases are many and contemplate anything from inline formatting to marking positions in the document. (The question of whether shortcodes remain the right tool is not ours to ask here!)

> It seems to me that this should be fine because shortcodes are converted to blocks, which can't be inline

Because even shortcodes that don't have a known associated block type will be picked up by the catch-all Shortcode block, we would be suddenly forcing all shortcodes to be block-like. One _could_ let the editor differentiate between shortcodes with a known block type (which can never be inline) and unknown shortcodes (which may or not be inline), but I'd honestly stay away from implementing that. It would introduce a lot of complexity for little benefit.

Instead, it would be more interesting if the block editor were smart enough to decide when a pasted shortcode is meant to be inline or not. If it is, the editor should abstain from trying to make it a block. Concretely, these are some behaviours to consider:

1. Start a new line (i.e. a blank Paragraph) and paste a shortcode → Add a block for that shortcode.
2. Place the caret in the middle of some text and paste a shortcode → Insert the shortcode text verbatim, no new blocks.
3. Place the caret at the end of a paragraph of text and paste a shortcode → …? (up to us to decide)

**Review by mcsf — CHANGES_REQUESTED** [review]

File: packages/block-editor/src/components/rich-text/index.js
Comment (nosolosw) [inline packages/block-editor/src/components/rich-text/index.js:98]: Is this too naive a RegExp?

**Comment by nosolosw** [comment]
@mcsf I've pushed a different approach that implements [your suggestion](https://github.com/WordPress/gutenberg/pull/21864#discussion_r416051965).

I think it's simpler if we left the use case 3 as it is at the moment, otherwise, we may be preventing the use of certain inline shortcodes that may behave like the end of block markers (by virtue of converting them blocks).

File: packages/block-editor/src/components/rich-text/index.js
Comment (mcsf) [reply packages/block-editor/src/components/rich-text/index.js:98]: You could grab [`regexp` from `@wordpress/shortcode`](https://github.com/WordPress/gutenberg/blob/68607318f23bfa371b7e1ec081dba88214d4ebd2/packages/shortcode/src/index.js#L155-L162) and [test it](https://github.com/WordPress/gutenberg/blob/68607318f23bfa371b7e1ec081dba88214d4ebd2/packages/blocks/src/api/raw-handling/shortcode-converter.js#L32), if you want the battle-tested stuff. :)

**Comment by mcsf** [comment]
> I think it's simpler if we left the use case 3 as it is at the moment, otherwise, we may be preventing the use of certain inline shortcodes that may behave like the end of block markers (by virtue of converting them blocks).

Yeah, this makes sense.

File: packages/block-editor/src/components/rich-text/index.js
Comment (nosolosw) [reply packages/block-editor/src/components/rich-text/index.js]: nice! 8cf006b

**Review by mcsf — APPROVED** [review]
Looks good now, thanks for fixing!

File: packages/block-editor/src/components/rich-text/index.js
Comment (ellatrix) [inline packages/block-editor/src/components/rich-text/index.js:42]: I didn't see this code at the time of approval. I feel like this belongs in the paste handler together with the rest of the short code logic?

Comment (nosolosw) [reply packages/block-editor/src/components/rich-text/index.js:42]: I think code may have changed since your approval and Miguel's. I didn't think of pinging you for your re-approval after Miguel's requested changes, sorry.

However, I don't know how we can move this logic to the paste handler. The trick here is to switch to `BLOCKS` mode if 1) we're in an empty line and 2) the pasted text is a shortcode. This logic requires knowledge of the rich text value and the pasted content. My thinking was that having this on the rich-text component would make sense as we already do the same for URLs. How does this look? Happy to refactor if another approach is preferred.

Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js:42]: Also the dependencies are missing in the package file

Comment (aduth) [reply packages/block-editor/src/components/rich-text/index.js:42]: > Also the dependencies are missing in the package file

Aside: I thought we had an ESLint rule which verified that any referenced dependencies should be listed in `package.json` 🤔

Comment (nosolosw) [reply packages/block-editor/src/components/rich-text/index.js:42]: Ouch :disappointed: Here's the fix for the missing dependency https://github.com/WordPress/gutenberg/pull/22086

Not sure I've done this before so I may have missed something.

Comment (nosolosw) [reply packages/block-editor/src/components/rich-text/index.js:42]: Do we need to make a package release due to this? I guess we do unless we can wait until next week's release. Not sure how you managed previous precedents if any, but happy to do it myself as a future reminder of this kind of thing.

Comment (aduth) [reply packages/block-editor/src/components/rich-text/index.js:42]: >Do we need to make a package release due to this?

I think as long as these updates haven't been included in a packages release, there's still time to correct the mistake without needing to do any extra publishing.

As best I can tell, the latest version doesn't include this line of code:

https://unpkg.com/browse/@wordpress/block-editor@3.10.0/build-module/components/rich-text/index.js

Comment (nosolosw) [reply packages/block-editor/src/components/rich-text/index.js:42]: oh, nice

Comment (aduth) [reply packages/block-editor/src/components/rich-text/index.js:42]: > > Also the dependencies are missing in the package file
> 
> Aside: I thought we had an ESLint rule which verified that any referenced dependencies should be listed in `package.json` 🤔

Something must not be working correctly, but we do define `import/no-extraneous-dependencies` as a rule, which is intended to enforce this:

https://github.com/WordPress/gutenberg/blob/9db5982eda68e8f923e60792242702c8a5960822/.eslintrc.js#L133

>Forbid the import of external modules that are not declared in the package.json's dependencies, devDependencies, optionalDependencies, peerDependencies, or bundledDependencies. 

https://github.com/benmosher/eslint-plugin-import/blob/HEAD/docs/rules/no-extraneous-dependencies.md

Comment (aduth) [reply packages/block-editor/src/components/rich-text/index.js:42]: Very strange indeed. It fails locally!

```
 ⇒ npm run lint-js                                                        

> gutenberg@8.0.0 lint-js /Users/andrew/Documents/Code/gutenberg
> wp-scripts lint-js


/Users/andrew/Documents/Code/gutenberg/packages/block-editor/src/components/rich-text/index.js
  42:1  error  '@wordpress/shortcode' should be listed in the project's dependencies. Run 'npm i -S @wordpress/shortcode' to add it  import/no-extraneous-dependencies

✖ 1 problem (1 error, 0 warnings)
```

Comment (nosolosw) [reply packages/block-editor/src/components/rich-text/index.js:42]: The [Travis job](https://travis-ci.com/github/WordPress/gutenberg/jobs/324466082) for this PR didn't report anything neither for all subsequent PRs. Just looked at the logs and there aren't any messages about the failure either (it could be that the exit code was wrong for some reason but the message was there).

I can confirm that `npm run lint-js` and `npm run lint` do fail locally on `master` prior to the merge of https://github.com/WordPress/gutenberg/pull/22086 Puzzling.

Comment (mcsf) [reply packages/block-editor/src/components/rich-text/index.js:42]: > Aside: I thought we had an ESLint rule which verified that any referenced dependencies should be listed in `package.json` 🤔

Oh, that's unfortunate, I let my guard down too. 😞 

> I didn't see this code at the time of approval. I feel like this belongs in the paste handler together with the rest of the short code logic?

Perhaps, but there are a lot of symmetries with what we do with pasted URLs and embeds. Whatever we do, we should do for both shortcodes and embeds. Perhaps that means letting `pasteHandler` choose whether to force `mode: 'BLOCKS'` when initial `mode` is `'AUTO'`; perhaps that means exposing a utility like `inferMode` or `shouldForceBlocksMode` from `@wordpress/blocks` that RichText can use.

Comment (aduth) [reply packages/block-editor/src/components/rich-text/index.js:42]: > Very strange indeed. It fails locally!

After some more debugging, I think I've figured it out.

Fix (and debugging details) at: #22088

Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js:42]: Sorry, I don't mind the code, it's something we can always easily change. I was thinking we could also pass empty state to the raw handler. I just thought it might be nice to keep shortcode dependencies in the raw handler since it already depends on it and RichText doesn't. I just noticed the code when I saw the package error.

**Pull request merged**
